### PR TITLE
dag-cbor: cleanup, reference RFC 8949, include @ipld/dag-cbor, mark final

### PR DESCRIPTION
**Pull request by rvagg**

### CODE REVIEW

**Review by vmx — APPROVED** [review]

File: block-layer/codecs/dag-cbor.md
Comment (mvdan) [inline block-layer/codecs/dag-cbor.md:134]: TIL these can contain special markdown characters

File: block-layer/codecs/dag-cbor.md
Comment (mvdan) [inline block-layer/codecs/dag-cbor.md:85]: out of curiosity, do we have an issue about this for ipld-prime?

File: block-layer/codecs/dag-cbor.md
Comment (mvdan) [inline block-layer/codecs/dag-cbor.md:86]: do we want to change this, given the recent discussions on the topic? @warpfork

**Review by mvdan — APPROVED** [review]

File: block-layer/codecs/dag-cbor.md
Comment (ribasushi) [reply block-layer/codecs/dag-cbor.md:86]: same question, LGTM otherwise

**Review by ribasushi — APPROVED** [review]

File: block-layer/codecs/dag-cbor.md
Comment (mikeal) [reply block-layer/codecs/dag-cbor.md:86]: I think we should leave out statements about when/how strictness is applied. The spec should define the strictness and implementations that want to violate the strictness can document when and where they do so.

We should strive for an equilibrium in which strictness is assumed to be the norm and is explicitly called out when not applied.

Comment (rvagg) [reply block-layer/codecs/dag-cbor.md:86]: > strictness is assumed to be the norm and is explicitly called out when not applied

that's what this section is doing, are you suggesting I remove this?

File: block-layer/codecs/dag-cbor.md
Comment (warpfork) [inline block-layer/codecs/dag-cbor.md:52]: I still hold that map sorting is an antifeature and it's almost certainly a mistake to specify it with a "must".

We've got lots of data in the wild that doesn't do this already, and lots of reasoning why it's ecosystemically dubious even to try to force it in the future.  I've written https://gist.github.com/warpfork/9315eac9848fa2c2ab36174272a80d9b#reasons about this and still feel it holds in its entirety.

I strongly suspect the outcomes of strict map sorting will be: people are pushed to use CBOR (and not DAG-CBOR) for handling any legacy data; and/or we'll see the emergence of a DAG-CBOR-HARDER codec in the future which is identical to this spec in every way but is order-preserving.  Or: we'll simply see almost every implementation not enforce this rule (in which case, what was the point?).

I would much rather we move this to a section on canonicalization which makes sorting *recommend* when the user data *is known to lack a meaningful ordering*.

I'm also not going to hold the PR up on this, because the previous text also stated sorting, and this text just states it better.

File: block-layer/codecs/dag-cbor.md
Comment (warpfork) [reply block-layer/codecs/dag-cbor.md:86]: It seems we should probably remove (or at least put behind a flag) our support for these "numbers", yes.  (I have no idea what our code currently does with these, in detail.)

File: block-layer/codecs/dag-cbor.md
Comment (warpfork) [reply block-layer/codecs/dag-cbor.md:85]: Probably need to break this down into a couple things:

- The NaN/Infinite issue
- Float encoding size (I think we accept variable size, but only emit 64bit size, iirc)
- Map sorting (but [I have opinions about whether this is a good idea](https://github.com/ipld/specs/pull/353/files#r561683154))

File: block-layer/codecs/dag-cbor.md
Comment (rvagg) [reply block-layer/codecs/dag-cbor.md:134]: I'm only using foot-links here because they're already in this doc, otherwise I _hate_ this style of MD linking

File: block-layer/codecs/dag-cbor.md
Comment (rvagg) [reply block-layer/codecs/dag-cbor.md:86]: Sorry, I'm getting mixed signals in this thread. For clarity - refmt supports `NaN`, `Infinity` and `-Infinity` but I'm assuming this will change at some point in the future, perhaps when ipld-prime absorbs more refmt functionality and does its own CBOR. I don't view this as particularly urgent work and don't think these listed items should be viewed as marks against the libraries, just informational for users.

@mikeal @warpfork are you suggesting I remove comments about how the various libraries conform or don't conform from here? That would seem to be a negative for this document, they're informational for users. If you have suggestions for how to reframe the wording then let me know.

File: block-layer/codecs/dag-cbor.md
Comment (rvagg) [reply block-layer/codecs/dag-cbor.md:52]: OK, let me first acknowledge your overall objection here @warpfork, but this particular change to the text is only clarifying what the current status is, not changing the current status. It would be nice if we could find a way to split out this objection from this particular PR rather than hold up a clean-up with a structural objection. How can I adjust to help this happen? Do you want me to remove the change to "Final"?

I have thoughts on how we might inch toward a reconciliation on the views here, but I really don't want to bikeshed in this PR.

File: block-layer/codecs/dag-cbor.md
Comment (mvdan) [reply block-layer/codecs/dag-cbor.md:134]: Really? I actually find it quite tidy and nice for reading the "plaintext English" without having to render the doc.

File: block-layer/codecs/dag-cbor.md
Comment (warpfork) [reply block-layer/codecs/dag-cbor.md:52]: Yeah, if we leave the final off for the moment being, I'd be happy overall.

File: block-layer/codecs/dag-cbor.md
Comment (warpfork) [reply block-layer/codecs/dag-cbor.md:86]: I am happy about how this is currently phrased.  (Github's UI may have caused me and Mikeal tunnel vision earlier; some views "helpfully" cut off the line introducing this list stating that it's a list of exceptions to the rule.  I think we're all in agreement.)

**Comment by rvagg** [comment]
Turned it back to "Draft" before merging as per comment above.

**Pull request merged**
